nilearn/regions/tests/test_signal_extraction.py

```diff
@@ -17,14 +17,119 @@
 from nilearn._utils.exceptions import DimensionError
 from nilearn.image import get_data, new_img_like
 
+
 _TEST_DIM_ERROR_MSG = ("Input data has incompatible dimensionality: "
                        "Expected dimension is 3D and you provided "
                        "a 4D image")
+_TEST_SHAPE_ERROR_MSG = "Images have incompatible shapes."
+_TEST_AFFINE_ERROR_MSG = "Images have different affine matrices."
 
+INF = 1000 * np.finfo(np.float32).eps
 
-def test_generate_regions_ts():
-    """Minimal testing of generate_regions_ts()"""
 
+def test__check_shape_and_affine_compatibility_smoke():
+    """Ensure correct behaviour for valid data."""
+    shape = (8, 9, 10)
+    affine = np.eye(4)
+
+    labels_img = generate_labeled_regions(shape, 7, affine=affine)
+
+    target_img = nibabel.Nifti1Image(np.zeros(shape), affine)
+
+    with pytest.warns(None):
+        signal_extraction._check_shape_and_affine_compatibility(
+            labels_img, target_img)
+
+
+def test__check_shape_and_affine_compatibility_error_shape():
+
+    shape = (8, 9, 10)
+    affine = np.eye(4)
+    labels_img = generate_labeled_regions(shape, 7, affine=affine)
+
+    test_shape = (8, 9, 5)
+    target_img = nibabel.Nifti1Image(np.zeros(test_shape), affine)
+
+    with pytest.raises(ValueError, match=_TEST_SHAPE_ERROR_MSG):
+        signal_extraction._check_shape_and_affine_compatibility(
+            labels_img, target_img)
+
+
+def test__check_shape_and_affine_compatibility_error_affine():
+
+    shape = (8, 9, 10)
+    affine = np.eye(4)
+    labels_img = generate_labeled_regions(shape, 7, affine=affine)
```

Comment:
It looks like this is the same image for multiple tests. Could you turn it into a fixture to avoid duplication ?

Reply:
hum not so sure I would do that.

As they are the tests make it quite clear when the affine and shape are different between the 2 images.

Further refactoring would obfuscate that and would force to go and check in the fixture what the values of the affine and the shapes are for one of the image.

Reply:
BTW there were some tests duplicate, so I could remove the entire tests which reduces the number of repeats of the set up.

Reply:
Fair enough, let's keep them as is then. 👍

Reply:
Will have one final go to try to see if we can keep things clear, readable and succinct with a test parametrization.